Add support for auto-generating multiple RCs for the current version if a previous RC exists

fullofcaffeine opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description

Corresponding issue: https://github.com/WordPress/gutenberg/issues/25970

**Latest update and test results here:** https://github.com/WordPress/gutenberg/pull/25971#issuecomment-708797141.

Add support for auto-generating multiple RCs or the current version if a previous RC exist.

If a previous RC exists, it will generate a new RC, in sequence. If not, it will fallback to the regular logic of generating an RC1 for the `nextMajor` version, as well as creating the new `releaseBranch`.


## How has this been tested?

I've run it locally and changed the config to point to my fork of GB. 

To test it out, change the values of https://github.com/WordPress/gutenberg/blob/master/bin/plugin/config.js to point to a test fork of Gutenberg. [Example](https://gist.github.com/fullofcaffeine/2ed4ac9d8fa219d5072defdf9aeac31b).

1. First, run `./bin/plugin/cli.js rc` from Gitenberg's root dir. This assumes the last release is a stable non-RC one. The behaviour here should be the same as before the changes here.

1. Now, you can take note of where the script cloned the version of GB downloaded when you ran the script in 1), and then manually set this var to point to its path: https://github.com/fullofcaffeine/gutenberg/blob/add/support-multiple-rcs-release-tool/bin/plugin/commands/release.js#L719 (i.e: `const gitWorkingDirectory = '/tmp/59506893-74ff-4012-9486-36a0c7ca2869';`)

1. It will then release RC1 for the next major version;

1. Run the script again and it should pick up the current version and will automatically increment it to RC2. 

1. Run `./bin/plugin/cli.js stable` to release a stable version.

1. Run it again to release a point release.



## Types of changes

New feature (non-breaking change which adds functionality) 


## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

fullofcaffeine committed:
Add support for auto-generating multiple RCs or the current version if a previous RC exists

If a previous RC exists, it will generate a new RC, in sequence. If not, it will fallback to the regular logic of generating a RC1 for the `nextMajor` version, as well as creating the new `releaseBranch`.

ockham commented on bin/plugin/commands/release.js:
```diff
@@ -71,34 +71,64 @@ async function runReleaseBranchCreationStep(
 	await runStep( 'Creating the release branch', abortMessage, async () => {
+		let parsedMajorVersion;
+
+		const getComponentsOfRC = ( ver ) => ver.match( /(.+-rc\.)([0-9]+)/ );
```
I think we can use `semver.parse()` instead, and inspect the resulting `prerelease` attribute. Something like

```js
const packageVersion = semver.parse( packageJson.version );
const { prerelease } = packageVersion;
```

ockham replied:
Example:

```
const version = semver.parse( '9.1.2-rc.1' );
console.log( version.prerelease );
```

gives `[ 'rc', 1 ]` :tada:

fullofcaffeine replied:
Ah, good one! I'll change 👍🏻

ockham commented on bin/plugin/commands/release.js:
```diff
@@ -71,34 +71,64 @@ async function runReleaseBranchCreationStep(
-		const nextMajor = getNextMajorVersion( packageJson.version );
-		const parsedMajorVersion = semver.parse( nextMajor );
+		const rcComponents = getComponentsOfRC( packageJson.version );
+		const isRC = !! rcComponents;
```
While `isRC` is accurate naming, it's prone to confusion with the (user-supplied) `isRC` bool flag in `releasePlugin()`. Let's maybe rename this one. `packageVersionIsRC` maybe?

ockham replied:
or `isPackageVersionRC`? That's more predicate-style.

fullofcaffeine replied:
Ohh, I missed that bool flag. I think `isPackageVersionRC` is a good name 👍🏻

ockham commented on bin/plugin/commands/release.js:
```diff
@@ -71,34 +71,64 @@ async function runReleaseBranchCreationStep(
+		const rcComponents = getComponentsOfRC( packageJson.version );
+		const isRC = !! rcComponents;
+
+		// if the last tag is an RC then we can assume that we should generate a new rc with the same version number
```
:+1: :100:

ockham commented on bin/plugin/commands/release.js:
```diff
@@ -71,34 +71,64 @@ async function runReleaseBranchCreationStep(
+
+		// if the last tag is an RC then we can assume that we should generate a new rc with the same version number
+		if ( isRC ) {
+			const [ , current, rc ] = rcComponents;
+			const nextRC = Number( rc ) + 1;
+
+			version = `${ current }${ nextRC }`;
+			versionLabel = `${ current.split( '-' )[ 0 ] } RC${ nextRC }`;
+			parsedMajorVersion = semver.parse( packageJson.version );
```
We can use `semver.inc()` for the version arithmetics here:

```js
version = semver.inc( packageVersion, 'prerelease' );
```

fullofcaffeine replied:
TIL 👍🏻

ockham reviewed (commented):
Left a few notes. Looking pretty good overall!

ockham commented:
We should also update this line: https://github.com/WordPress/gutenberg/blob/d4a74cca63617405b2bf2680ef8011dfb87dd367/bin/plugin/cli.js#L34

fullofcaffeine committed:
Use semver to detect if its prerelease and to incremenet RCs

Improve variable naming, simplify the code.

fullofcaffeine committed:
Update the description for the release-plugin-rc command to match the new behavior of releaseing multiple RCs for the same version

ockham commented on bin/plugin/cli.js:
```diff
@@ -31,7 +31,7 @@ program
 	.alias( 'rc' )
 	.description(
-		'Release an RC version of the plugin (supports only rc.1 for now)'
+		'Release an RC version of the plugin (will increment the RC number if one already exists for the last version)'
```
We can probably drop the note in parens altogether, I think it's kinda implied, and we're doing the same for stable versions where we also don't mention the increment.

```suggestion
		'Release an RC version of the plugin'
```

fullofcaffeine committed:
Simplify the command description

Co-authored-by: Bernie Reiter <ockham@raz.or.at>

ockham commented on bin/plugin/commands/release.js:
```diff
@@ -69,40 +69,59 @@ async function runReleaseBranchCreationStep(
 
-		version = nextMajor + '-rc.1';
+		// if the last tag is an RC then we can assume that we should generate a new rc with the same version number
+		if ( isPackageVersionRC ) {
+			parsedVersion = semver.parse(
+				semver.inc( packageVersion.version, 'prerelease', 'rc' )
+			);
+		} else {
+			const nextMajor = getNextMajorVersion( packageJson.version );
+			parsedVersion = semver.parse( nextMajor + '-rc.1' );
+		}
+
+		versionLabel = parsedVersion.raw.replace( /\-rc\.([0-9]+)/, ' RC$1' );
```
Let's maybe use a (non-parsed) `version` as the variable, since we need that more often than the parsed one:

```suggestion
		if ( isPackageVersionRC ) {
			version = semver.inc( packageVersion.version, 'prerelease', 'rc' );
		} else {
			const nextMajor = getNextMajorVersion( packageJson.version );
			version = nextMajor + '-rc.1';
		}

		versionLabel = version.replace( /\-rc\.([0-9]+)/, ' RC$1' );
```

fullofcaffeine committed:
Avoid semver-parsing the version multiple times and parse just before it is needed

We need to parse twice, first the version from the package to be able to use the `prerelease` property so we can decide how to proceed. Then after we know if the current version is an RC or not, we parse the next version so that the releaseBranch can be defined.

ockham committed:
Whitespace

ockham reviewed (approved):
LGTM, and tests well (as far as I could test).

Let's wait for tests to go green, and ideally someone who's more familiar with these tools to approve :+1: 

cc/ @youknowriad @gziolo @mcsf

fullofcaffeine committed:
Force-push to the release branch to avoid ref errors in sub-sequent RC releases

Since the release is cherry-picked into master, the SHA changes, thus, pushing to the release branch is not possible anymore without force-pushing.

fullofcaffeine commented on bin/plugin/commands/release.js:
```diff
@@ -390,9 +387,11 @@ async function runPushGitChangesStep(
 				true,
 				abortMessage
 			);
+			// We need to force-push because we are effectivelly rebasing from master
```
Since the release is cherry-picked into master, the SHA changes, thus, pushing to the release branch is not possible anymore without force-pushing. 

This seems to work well, but I don't know if there might be unforeseen side-effects to force-pushing to the release branch. I'd love some feedback here.

youknowriad replied:
I don't like this. we shouldn't be force pushing to master IMO, we could potentially lose merges that happened in the meantime.

mcsf replied:
At the very least, this should be [`--force-with-lease`](https://git-scm.com/docs/git-push).

(Btw, everyone should use that option in their daily Git interactions — I recommend adding aliases to Git like ` fpush = push --force-with-lease` or `fp` or `pf` or whatever.)

ockham replied:
Our ideal scenario here would be to have all RC commits (alongside potential other, cherry-picked, ones) on the release branch, no? So we need to ensure that we start off from the point where we "left off" with the previous RC. I'm wondering what's currently happening that gets in the way of that?

Could it be that the `git.checkoutLocalBranch` call in `runReleaseBranchCreationStep` should only be called if the branch didn't exist before, and that we need to run `checkoutRemoteBranch` otherwise?

ockham replied:
> Since the release commit is cherry-picked into master, the SHA changes, thus, pushing to the release branch is not possible anymore without force-pushing in the case of subsequent RCs.

Are you sure that's the reason? :thinking: Wouldn't we run into the same problem when releasing a stable version after an RC?

ockham replied:
I'm starting to think it would make sense to change the criterion for the ternary that runs either `runReleaseBranchCreationStep` or `runReleaseBranchCheckoutStep` from `isRC` to `isPackageVersionRC`, since it captures better what we're trying to do here:

- If the `package.json` version is an RC, then we can assume that a release branch has been created.
- If it doesn't, we're at the start of a new release cycle and need to create that branch.

This will mean moving some of the version computing logic around, but I think it'll fall in place quite naturally.

youknowriad replied:
@ockham that makes sense since second/third RCs will just just use the existing branch with cherrypicked commits

fullofcaffeine replied:
> I don't like this. we shouldn't be force pushing to master IMO, we could potentially lose merges that happened in the meantime.

Note that the force push here is not to master, but to the `releaseBranch`.

fullofcaffeine committed:
Simplify jsdoc type

fullofcaffeine commented on bin/plugin/lib/git.js:
```diff
@@ -82,10 +82,20 @@ async function createLocalTag( gitWorkingDirectoryPath, tagName ) {
 	const simpleGit = SimpleGit( gitWorkingDirectoryPath );
-	await simpleGit.push( 'origin', branchName );
+
+	// TODO The type definitions for simple-git are wrong and will complain here.
```
As the comment states, there seems to be a mismatch between the TS typing declarations and the js version of simple-git. I tried passing the `options` array as the last arg as the type docs say, but it didn't work. 

After inspecting the actual code, I realized thet last param is called "then" and is unused. By checking the implementation, I saw that the right way to pass these options was to pass them as the first argument.

I'll look into it again tomorrow, but if anyone else more acquainted with the codebase has any insights, please let me know :)

fullofcaffeine replied:
I managed to fix this by passing the options as an object instead of an array (as the last argument). The interface here is not the most intuitive ¯\_(ツ)_/¯

fullofcaffeine committed:
Change options argument passed over to simple-git push to an object

This method has a very confusing signature. The comment in the source-code suggests you can pass all arguments as an options array as part of the 1st argument (remote) and that's what I was doing. However, it's not compatible with the type definitions. You can't pass an array of options as the last arguments as it will just swallow them silently. For the last argument to be recognized as git options, you should pass it as a JS object. If the options does not accept arguments, its value should be null.

Changing it to an object fixed the type errors too.

fullofcaffeine commented:
This seems to work well. I just tested twice in order to get RC1 and the subsequent RC2:

1. Output for RC1: https://gist.github.com/fullofcaffeine/848e637f16c0b856766185620d9574ee
2. Output for RC2: https://gist.github.com/fullofcaffeine/d0a2880ac430bc60d40d67547faee24c

youknowriad commented:
This is cool, aside the force push which I think is very risky,  this is looking good.

The pre-release is normal, we mark all "RCs" as "pre-releases" explicitly in the script.

mcsf commented on bin/plugin/commands/release.js:
```diff
@@ -73,32 +73,48 @@ async function runReleaseBranchCreationStep(
+			);
 
-		// Creating the release branch
+			log(
+				'>> The local release branch ' +
+					formats.success( releaseBranch ) +
+					' has been successfully created.'
+			);
```
Shouldn't this remain after the call to `git.createLocalBranch`?

fullofcaffeine replied:
Good point. The call to `createLocalBranch` has been moved down becuase I found out that was also needed for the RC scenario, as it essentially checkout th branch. It doesn't make sense indeed to show a confirmation message before the actual thing happened :)

ockham committed:
Re-arrange version arithmetics

ockham committed:
Add back exports :rolleyes:

fullofcaffeine committed:
Re-arrange the version versioning logic a bit

Be more explicit, fix case where stable -> stable would try to create a branch instead of checking out the existing for the version.

fullofcaffeine commented:
Thanks @ockham for lending your hand today and to help me to understand the underlying issue (that led me to try the force-push approach*) 🙇🏻 

I had to re-arrange the logic slithgly again as it was failing just when trying to update from stable to stable. I tried to simplify it a bit, too. 

I did another round of tests (using my own fork of GB, see config [here](https://gist.github.com/fullofcaffeine/2ed4ac9d8fa219d5072defdf9aeac31b)) and it now seems to work well (no more force-push needed). Here are the scenarios and their respective results:

|  command               | package version | new version | observations |
| ------------- | ------------- | --- | --- |
| rc | `9.1.1`  | `9.2.0-rc.1` | Uses [this](https://github.com/fullofcaffeine/gutenberg/blob/add/support-multiple-rcs-release-tool/bin/plugin/lib/version.js#L13) function to generate the next version |
| rc  | `9.2.0-rc.1`  | `9.2.0-rc.2` | increments the RC ad infinitum until a stable version is released |
| stable | `9.2.0-rc.2` | `9.2.0` | increments the patch version, effectively removing the RC portion |
| stable | `9.2.0` | `9.2.1` | increments the patch version ad infinitum until we release another RC* |

_Note that I didn't fully finish the whole process when running the `stable` command, I replied 'N' when it got to the SVN part, but I don't think it's needed to consider these tests as succesful, as all the relevant steps have been run prior to it._

*This effectively makes the `rc` command mandatory for new [minor/major](https://github.com/fullofcaffeine/gutenberg/blob/add/support-multiple-rcs-release-tool/bin/plugin/lib/version.js#L13) versions.

_*Basically, we were trying to create a branch and push in place of checking out the existing branch for the scenarios that created a release in the same major/minor version. We had to adjust the logic to be aware of that, too._

ockham commented on bin/plugin/commands/release.js:
```diff
@@ -189,7 +146,6 @@ async function runReleaseBranchCheckoutStep(
 
 	return {
 		version,
```
Let's drop `version` from the returned object -- we're expecting it as a function arg now and aren't modifying it, so there's no point really to return it.

fullofcaffeine replied:
You're right, I removed versionLabel for this reason and this slipped through the cracks.

ockham replied:
Thanks for fixing!

ockham commented on bin/plugin/commands/release.js:
```diff
@@ -60,123 +60,80 @@ async function runSvnRepositoryCheckoutStep( abortMessage ) {
 	} );
 
 	return {
 		version,
```
Same as https://github.com/WordPress/gutenberg/pull/25971/commits/069f5f8cc1c0e6b3d981ad74744f8b9ee81b1fd1#r505636588

ockham commented on bin/plugin/commands/release.js:
```diff
@@ -720,21 +678,39 @@ async function releasePlugin( isRC = true ) {
+		);
+	}
+
+	const { version, releaseBranch } = versionInfo;
```
See https://github.com/WordPress/gutenberg/pull/25971/commits/069f5f8cc1c0e6b3d981ad74744f8b9ee81b1fd1#r505636588 -- let's not obtain `version` from the return value of `runReleaseBranchCreationStep`/`runReleaseBranchCheckoutStep`, but simply set a `version` var inside of our `if`/`else` construct, and use that both as an arg _to_ those functions, _and_ here.

We might consider doing the same for `releaseBranch` -- we might just compute that value in this function now, and pass it to `runReleaseBranchCreationStep`/`runReleaseBranchCheckoutStep`.

fullofcaffeine committed:
Remove redundant version return and DRY things a bit more

ockham reviewed (approved):
Thanks, this is testing well for me now :+1:

ockham commented:
@youknowriad @mcsf Could we get your final blessing please? :pray:

youknowriad commented:
Ship it, thanks for your work here :)
